server/src/main/java/org/opensearch/gateway/remote/RemoteGlobalMetadataManager.java

```diff
@@ -0,0 +1,334 @@
+    private final RemoteClusterStateBlobStore<TemplatesMetadata, RemoteTemplatesMetadata> templatesMetadataBlobStore;
+    private final RemoteClusterStateBlobStore<Custom, RemoteCustomMetadata> customMetadataBlobStore;
+    private final Compressor compressor;
+    private final NamedXContentRegistry namedXContentRegistry;
+
+    RemoteGlobalMetadataManager(
+        ClusterSettings clusterSettings,
+        RemoteWritableEntityStore<Metadata, RemoteGlobalMetadata> globalMetadataBlobStore,
+        RemoteClusterStateBlobStore<CoordinationMetadata, RemoteCoordinationMetadata> coordinationMetadataBlobStore,
+        RemoteClusterStateBlobStore<Settings, RemotePersistentSettingsMetadata> persistentSettingsBlobStore,
+        RemoteClusterStateBlobStore<TemplatesMetadata, RemoteTemplatesMetadata> templatesMetadataBlobStore,
+        RemoteClusterStateBlobStore<Custom, RemoteCustomMetadata> customMetadataBlobStore,
+        Compressor compressor,
+        NamedXContentRegistry namedXContentRegistry
+    ) {
+        this.globalMetadataUploadTimeout = clusterSettings.get(GLOBAL_METADATA_UPLOAD_TIMEOUT_SETTING);
+        this.compressor = compressor;
+        this.globalMetadataBlobStore = globalMetadataBlobStore;
+        this.namedXContentRegistry = namedXContentRegistry;
+        this.coordinationMetadataBlobStore = coordinationMetadataBlobStore;
+        this.persistentSettingsBlobStore = persistentSettingsBlobStore;
+        this.templatesMetadataBlobStore = templatesMetadataBlobStore;
+        this.customMetadataBlobStore = customMetadataBlobStore;
+        clusterSettings.addSettingsUpdateConsumer(GLOBAL_METADATA_UPLOAD_TIMEOUT_SETTING, this::setGlobalMetadataUploadTimeout);
+    }
+
+    /**
+     * Allows async upload of Metadata components to remote
+     */
+    CheckedRunnable<IOException> getAsyncMetadataWriteAction(
+        Object objectToUpload,
+        long metadataVersion,
+        String clusterUUID,
+        LatchedActionListener<ClusterMetadataManifest.UploadedMetadata> latchedActionListener,
+        String customType
+    ) {
+        if (objectToUpload instanceof CoordinationMetadata) {
+            RemoteCoordinationMetadata remoteCoordinationMetadata = new RemoteCoordinationMetadata(
+                (CoordinationMetadata) objectToUpload,
+                metadataVersion,
+                clusterUUID,
+                compressor,
+                namedXContentRegistry
+            );
+            return () -> coordinationMetadataBlobStore.writeAsync(
+                remoteCoordinationMetadata,
+                getActionListener(remoteCoordinationMetadata, latchedActionListener)
+            );
+        } else if (objectToUpload instanceof Settings) {
+            RemotePersistentSettingsMetadata remotePersistentSettingsMetadata = new RemotePersistentSettingsMetadata(
+                (Settings) objectToUpload,
+                metadataVersion,
+                clusterUUID,
+                compressor,
```

Comment:
Make the consumers of this method pass the `Remote`Metadata instance and avoid if-else checks